### PR TITLE
Enhance metric support

### PR DESCRIPTION
**Pull request by anbrsap**
### Description

- Refactor metric support code
- Embed workqueue metrics
- Embed client-go rest client metrics
- Use consistent metric names
- Add metrics for retry loops
- Deprecate metric `steward_pipelinerun_update_seconds`
- Add metrics reference documentation

### Submitter checklist

- [x] Change has been tested (on a back-end cluster)
- N/A (If applicable) Jira backlog item ID added to the PR title and the [changelog.yaml] entry
- [x] [changelog.yaml] entry with upgrade notes is prepared and appropriate for the audience affected by the change (users or developer, depending on the change)
- [x] Semantic version diffed against [last release][releases] and updated accordingly. In this project the version has to be maintained here:
    - [/charts/steward/Chart.yaml](https://github.com/SAP/stewardci-core/blob/master/charts/steward/Chart.yaml) (`version` and `appVersion`)

In case dependencies have been updated:
- N/A Links to external changelogs, since the last release of our component, added to the [changelog.yaml] entry (description).
- N/A Changelogs read thoroughly, potential impact described, upgrade notes prepared (if necessary)
- N/A Check if dependency updates affect our semantic version increment.

### Reviewer checklist

Before the changes are marked as `ready-for-merge`: 

- [x] There is at least 1 approval for the pull request and no outstanding requests for change
- [x] All voter checks have passed
- [x] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [x] The Pull Request title is understandable and reflects the changes well
- [ ] The Pull Request description is understandable and well documented
- [ ] [changelog.yaml] entry for this Pull Request has been added
    - [ ] Changelog entry contains all required information
    - [ ] 'Upgrade notes' are documented in changelog.yaml (if required)

[changelog.yaml]: https://github.com/SAP/stewardci-core/changelog.yaml
[releases]: https://github.com/SAP/stewardci-core/releases

### CODE REVIEW

File: pkg/metrics/retries.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+func (m *retriesMetric) init() {
+	m.initOnlyOnce.Do(func() {
+		m.countMetric = prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Subsystem: Subsystem,
+				Name:      "retry_count",
+				Help:      "Generic metric for retry loops collecting the number of retries performed for retried operations.",
+				// TODO review buckets
+				Buckets: prometheus.ExponentialBuckets(1, 2, 32),
```
Comment (rinckm) [inline pkg/metrics/retries.go:45]:
```suggestion
				Buckets: append(prometheus.LinearBuckets (0, 1, 10),prometheus.LinearBuckets (10, 5, 10),
```

File: pkg/metrics/workqueue/naming.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+func (f NameProviderFunc) GetSubsystemFor(queueName string) (string, bool) {
+	return f(queueName)
+}
+
+// nameProviders is a composite name provider that queries all registered
+// name provides in the order of registration. The first positive result
+// will be returned.
+type nameProviders struct {
+	lock     sync.Mutex
+	children []NameProvider
```
Comment (rinckm) [inline pkg/metrics/workqueue/naming.go:37]: Why are you not using a map with queue name as key?

Comment (anbrsap) [reply pkg/metrics/workqueue/naming.go:37]: First, this share package should not know it's consumers. This is why the naming must come from the packages where the respective queue belongs to.

Second, a name provider may also use pattern matching. A static name mapping would not allow that. Having a simple interface that is queried for each name looks reasonable to me.

File: pkg/metrics/registry.go [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	registry = prometheus.NewPedanticRegistry()
+)
+
+// Registerer returns the registerer for metrics that should be exported
+// by the metrics server
+func Registerer() prometheus.Registerer {
```
Comment (rinckm) [inline pkg/metrics/registry.go:13]: Only MustRegister is used. We could also implement it here:

func MustRegister(cs ...Collector) {
  registry.MustRegister(cs)
}

Comment (anbrsap) [reply pkg/metrics/registry.go:13]: > Only MustRegister is used.

Just coincidence :-) There's a concise interface for registries, `prometheus.Registerer`. Why should I artificially limit it?

**Review by rinckm — APPROVED** [review]
Please check https://github.com/SAP/stewardci-core/pull/282#discussion_r753001981

File: pkg/metrics/retries.go (hunk 1 above)
Comment (anbrsap) [reply pkg/metrics/retries.go:45]: Given the fact that this is a _generic_ metric, there might measured retry loops that typically go beyond 75 retries. On the other hand we should not have too many buckets to not occupy too much space in the metrics store.

Here's my proposal:

```suggestion
				// TODO review buckets
				Buckets: append(prometheus.LinearBuckets (0, 1, 10), prometheus.ExponentialBuckets(10, 2, 11)),
```

This gives the following bucket upper bounds:

1
2
3
4
5
6
7
8
9
10
20
40
80
160
320
640
1280
2560
5120
10240

**Pull request merged**